Add web.dev/monitor-total-page-memory-usage

ulan opened the pull request:
This fixes #2475

jpmedley commented:
Is this still called `measureMemory`. I don't see that name in the latest IDL, but I do see a `memory` property?

@ulan

jpmedley commented:
I just notice this: "Currently web developers resort to the non-standard performance.memory API that is used in 20% of page loads."

Better question: is this actually shipping as planned in 83?

ulan commented:
@jpmedley do you mean performance.idl in Chrome?

measureMemory is there: https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/timing/performance.idl?l=70&rcl=8f5be71f0619c86fc24a7bd9d6d1b2151c869b93

The origin trial is active in 82/83.

jpmedley commented:
That answers my question. For some reason, I couldn't find it yesterday.

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+---
+title: Monitor your web page's total memory usage with performance.measureMemory()
```
```suggestion
title: Monitor your web page's total memory usage with `performance.measureMemory()`
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+It is easy to introduce one by forgetting to unregister an event listener, by
+accidentally capturing objects from an iframe, by not closing a worker, by
+accumulating objects in arrays, and so on. If a web page has memory leaks,
+then its memory usage grows over time and the web page appears as slow and
```
```suggestion
then its memory usage grows over time and the web page appears slow and
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+bloated to the users.
+
+The first step in solving this problem is measuring it. The new
+[performance.measureMemory API][explainer] allows developers to
```
```suggestion
[`performance.measureMemory()` API][explainer] allows developers to
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+measure memory usage of their web pages in production and thus detect memory
+leaks that slip through local testing.
+
+## How is performance.measureMemory() different from the legacy performance.memory API? {: #legacy-api }
```
```suggestion
## How is `performance.measureMemory()` different from the legacy `performance.memory` API? {: #legacy-api }
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+  - post
+  - memory
+  - javascript
+  - capabilities
```
```suggestion
```

This API doesn't seem to be part of the Capabilities project: https://docs.google.com/spreadsheets/d/1de0ZYDOcafNXXwMcg4EZhT0346QM-QFvZfoD8ZffHeA/edit#gid=557099940

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+</table>
+</div>
+
+## Using performance.measureMemory() {: use }
```
```suggestion
## Using `performance.measureMemory()` {: use }
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+
+### Feature detection
+
+The `performance.measureMemory` function may fail with a
```
```suggestion
The `performance.measureMemory()` function may fail with a
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+([demo][demo], [source][demo-source]).
+
+First, define a function that schedules the next memory measurement using
+setTimeout with a randomized interval. The function should be called after
```
```suggestion
`setTimeout()` with a randomized interval. The function should be called after
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
````diff
@@ -0,0 +1,357 @@
+}
+```
+
+The `measurementInterval` function computes a random interval in milliseconds
````
```suggestion
The `measurementInterval()` function computes a random interval in milliseconds
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
````diff
@@ -0,0 +1,357 @@
+```
+
+The `measurementInterval` function computes a random interval in milliseconds
+such that on average there is one measurement every five minutes. See [this
+article][math] if you are interested in the math behind the function.
````
```suggestion
such that on average there is one measurement every five minutes. See [Exponential
distribution][math] if you are interested in the math behind the function.
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,357 @@
+
+Are you planning to use `performance.measureMemory()`? Your public support
+helps the Chrome team prioritize features and shows other browser vendors how
+critical it is to support them. Send a tweet to @ChromiumDev and let us know
```
```suggestion
critical it is to support them. Send a tweet to [@ChromiumDev](https://twitter.com/chromiumdev) and let us know
```

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,356 @@
+date: 2020-04-07
+authors:
+  - ulan
+hero: hero.jpg
```
@ulan please optimize this with [squoosh](squoosh.app) so that we keep the size of the repository down

ulan replied:
@kaycebasques thanks for the tip, reduces the size to 116K

kaycebasques commented on src/site/content/en/blog/performance-measure-memory/index.md:
```diff
@@ -0,0 +1,356 @@
+---
```
@ulan please rename this directory to `monitor-total-memory-usage` or `performance-measurememory`

The first option is more aligned with our policy that the page URL should match the page title, the second option is more aligned with the actual name of the API (having a dash between `measure` and `memory` doesn't make sense to me)

ulan replied:
Renamed to `monitor-total-page-memory-usage`

ulan commented:
@kaycebasques thanks a lot for the corrections and comments!

I addressed the comments. Do we need to bump the date (2020-04-07) to the current date?

kaycebasques commented on src/site/content/en/blog/monitor-total-page-memory-usage/index.md:
```diff
@@ -0,0 +1,356 @@
+  Learn how to measure memory usage of your web page in production to detect regressions.
+description: >
+  Learn how to measure memory usage of your web page in production to detect regressions.
+date: 2020-04-07
```
```suggestion
date: 2020-04-13
```

kaycebasques commented:
Woohoo thanks @ulan. I'll merge the PR once the CI tests pass. I bumped the `date` field to today's date. The page will go live shortly after merging.

ulan commented:
@kaycebasques awesome! Thank you very much!
